petsc: fixup petscvariable; add external package superlu_dist

qbisi opened the pull request:
superlu_dist is needed in fullCheck of fenics-dolfinx.
since superlu_dist is supported on both linux and darwin platform now, i make it a commonDep when mpiSupport is enabled.

fixup petscvariable is needed when pyop2 in firedrakeProject lookup petscvariable for how to compile the generated c code.

both fenics and firedrake Project is going to publish a release by the end of April. I will make a Pr when release is available.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

qbisi committed:
petsc: add compile and link flags in lib/petsc/conf/petscvariables

qbisi committed:
petsc: add external package superlu_dist

NickCao commented on pkgs/by-name/pe/petsc/package.nix:
```diff
@@ -14,6 +14,8 @@
 
   # Build options
+  openblas,
+  blasProvider ? openblas,
```
I don't think blasProvider should be an option in downstream packages.

NickCao commented on pkgs/by-name/pe/petsc/package.nix:
```diff
@@ -208,6 +213,26 @@ stdenv.mkDerivation (finalAttrs: {
 
   enableParallelBuilding = true;
 
+  # add compile and link flags in lib/petsc/conf/petscvariables
```
Why's this not handled by upstream build system?

qbisi committed:
petsc: fix configureFlags for blaslapack

qbisi committed:
petsc: add test for varient built with mpich

mpich is not yet in pkgs/by-name and requires an explicit extra argument
`ch4backend` when used with callPackage, unlike openmpi. As a result, we
cannot override `mpi` in the petscPackages scope using the form
`callPackage mpi.override { }`.

qbisi commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review-gha`](https://github.com/Defelo/nixpkgs-review-gha)

Command: `nixpkgs-review pr 400357`

Logs: https://github.com/qbisi/nixpkgs-review-gha/actions/runs/14779012472


---
### `x86_64-linux` (sandbox = true)
<details>
  <summary>:white_check_mark: 13 packages built:</summary>
  <ul>
    <li>getdp</li>
    <li>petsc</li>
    <li>pflotran</li>
    <li>precice</li>
    <li>python312Packages.petsc4py</li>
    <li>python312Packages.pyprecice</li>
    <li>python312Packages.pyprecice.dist</li>
    <li>python312Packages.slepc4py</li>
    <li>python313Packages.petsc4py</li>
    <li>python313Packages.pyprecice</li>
    <li>python313Packages.pyprecice.dist</li>
    <li>python313Packages.slepc4py</li>
    <li>slepc</li>
  </ul>
</details>


---
### `aarch64-linux` (sandbox = true)
<details>
  <summary>:white_check_mark: 13 packages built:</summary>
  <ul>
    <li>getdp</li>
    <li>petsc</li>
    <li>pflotran</li>
    <li>precice</li>
    <li>python312Packages.petsc4py</li>
    <li>python312Packages.pyprecice</li>
    <li>python312Packages.pyprecice.dist</li>
    <li>python312Packages.slepc4py</li>
    <li>python313Packages.petsc4py</li>
    <li>python313Packages.pyprecice</li>
    <li>python313Packages.pyprecice.dist</li>
    <li>python313Packages.slepc4py</li>
    <li>slepc</li>
  </ul>
</details>


---
### `x86_64-darwin` (sandbox = true)
<details>
  <summary>:fast_forward: 12 packages marked as broken and skipped:</summary>
  <ul>
    <li>petsc</li>
    <li>pflotran</li>
    <li>precice</li>
    <li>python312Packages.petsc4py</li>
    <li>python312Packages.pyprecice</li>
    <li>python312Packages.pyprecice.dist</li>
    <li>python312Packages.slepc4py</li>
    <li>python313Packages.petsc4py</li>
    <li>python313Packages.pyprecice</li>
    <li>python313Packages.pyprecice.dist</li>
    <li>python313Packages.slepc4py</li>
    <li>slepc</li>
  </ul>
</details>


---
### `aarch64-darwin` (sandbox = true)
<details>
  <summary>:x: 12 packages failed to build:</summary>
  <ul>
    <li>petsc</li>
    <li>pflotran</li>
    <li>precice</li>
    <li>python312Packages.petsc4py</li>
    <li>python312Packages.pyprecice</li>
    <li>python312Packages.pyprecice.dist</li>
    <li>python312Packages.slepc4py</li>
    <li>python313Packages.petsc4py</li>
    <li>python313Packages.pyprecice</li>
    <li>python313Packages.pyprecice.dist</li>
    <li>python313Packages.slepc4py</li>
    <li>slepc</li>
  </ul>
</details>

qbisi committed:
petsc: add __darwinAllowLocalNetworking = true to support mpiCheck

qbisi commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review-gha`](https://github.com/Defelo/nixpkgs-review-gha)

Command: `nixpkgs-review pr 400357`

Logs: https://github.com/qbisi/nixpkgs-review-gha/actions/runs/14782116626


---
### `x86_64-linux` (sandbox = true)
<details>
  <summary>:white_check_mark: 13 packages built:</summary>
  <ul>
    <li>getdp</li>
    <li>petsc</li>
    <li>pflotran</li>
    <li>precice</li>
    <li>python312Packages.petsc4py</li>
    <li>python312Packages.pyprecice</li>
    <li>python312Packages.pyprecice.dist</li>
    <li>python312Packages.slepc4py</li>
    <li>python313Packages.petsc4py</li>
    <li>python313Packages.pyprecice</li>
    <li>python313Packages.pyprecice.dist</li>
    <li>python313Packages.slepc4py</li>
    <li>slepc</li>
  </ul>
</details>


---
### `aarch64-linux` (sandbox = true)
<details>
  <summary>:white_check_mark: 13 packages built:</summary>
  <ul>
    <li>getdp</li>
    <li>petsc</li>
    <li>pflotran</li>
    <li>precice</li>
    <li>python312Packages.petsc4py</li>
    <li>python312Packages.pyprecice</li>
    <li>python312Packages.pyprecice.dist</li>
    <li>python312Packages.slepc4py</li>
    <li>python313Packages.petsc4py</li>
    <li>python313Packages.pyprecice</li>
    <li>python313Packages.pyprecice.dist</li>
    <li>python313Packages.slepc4py</li>
    <li>slepc</li>
  </ul>
</details>


---
### `x86_64-darwin` (sandbox = true)
<details>
  <summary>:fast_forward: 12 packages marked as broken and skipped:</summary>
  <ul>
    <li>petsc</li>
    <li>pflotran</li>
    <li>precice</li>
    <li>python312Packages.petsc4py</li>
    <li>python312Packages.pyprecice</li>
    <li>python312Packages.pyprecice.dist</li>
    <li>python312Packages.slepc4py</li>
    <li>python313Packages.petsc4py</li>
    <li>python313Packages.pyprecice</li>
    <li>python313Packages.pyprecice.dist</li>
    <li>python313Packages.slepc4py</li>
    <li>slepc</li>
  </ul>
</details>


---
### `aarch64-darwin` (sandbox = true)
<details>
  <summary>:x: 3 packages failed to build:</summary>
  <ul>
    <li>python312Packages.slepc4py</li>
    <li>python313Packages.slepc4py</li>
    <li>slepc</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 9 packages built:</summary>
  <ul>
    <li>petsc</li>
    <li>pflotran</li>
    <li>precice</li>
    <li>python312Packages.petsc4py</li>
    <li>python312Packages.pyprecice</li>
    <li>python312Packages.pyprecice.dist</li>
    <li>python313Packages.petsc4py</li>
    <li>python313Packages.pyprecice</li>
    <li>python313Packages.pyprecice.dist</li>
  </ul>
</details>

qbisi committed:
slepc: add __darwinAllowLocalNetworking = true to support mpiCheck

qbisi commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review-gha`](https://github.com/Defelo/nixpkgs-review-gha)

Command: `nixpkgs-review pr 400357`

Logs: https://github.com/qbisi/nixpkgs-review-gha/actions/runs/14817256854


---
### `x86_64-linux` (sandbox = true)
<details>
  <summary>:white_check_mark: 13 packages built:</summary>
  <ul>
    <li>getdp</li>
    <li>petsc</li>
    <li>pflotran</li>
    <li>precice</li>
    <li>python312Packages.petsc4py</li>
    <li>python312Packages.pyprecice</li>
    <li>python312Packages.pyprecice.dist</li>
    <li>python312Packages.slepc4py</li>
    <li>python313Packages.petsc4py</li>
    <li>python313Packages.pyprecice</li>
    <li>python313Packages.pyprecice.dist</li>
    <li>python313Packages.slepc4py</li>
    <li>slepc</li>
  </ul>
</details>


---
### `aarch64-linux` (sandbox = true)
<details>
  <summary>:white_check_mark: 13 packages built:</summary>
  <ul>
    <li>getdp</li>
    <li>petsc</li>
    <li>pflotran</li>
    <li>precice</li>
    <li>python312Packages.petsc4py</li>
    <li>python312Packages.pyprecice</li>
    <li>python312Packages.pyprecice.dist</li>
    <li>python312Packages.slepc4py</li>
    <li>python313Packages.petsc4py</li>
    <li>python313Packages.pyprecice</li>
    <li>python313Packages.pyprecice.dist</li>
    <li>python313Packages.slepc4py</li>
    <li>slepc</li>
  </ul>
</details>


---
### `x86_64-darwin` (sandbox = true)
<details>
  <summary>:fast_forward: 12 packages marked as broken and skipped:</summary>
  <ul>
    <li>petsc</li>
    <li>pflotran</li>
    <li>precice</li>
    <li>python312Packages.petsc4py</li>
    <li>python312Packages.pyprecice</li>
    <li>python312Packages.pyprecice.dist</li>
    <li>python312Packages.slepc4py</li>
    <li>python313Packages.petsc4py</li>
    <li>python313Packages.pyprecice</li>
    <li>python313Packages.pyprecice.dist</li>
    <li>python313Packages.slepc4py</li>
    <li>slepc</li>
  </ul>
</details>


---
### `aarch64-darwin` (sandbox = true)
<details>
  <summary>:white_check_mark: 12 packages built:</summary>
  <ul>
    <li>petsc</li>
    <li>pflotran</li>
    <li>precice</li>
    <li>python312Packages.petsc4py</li>
    <li>python312Packages.pyprecice</li>
    <li>python312Packages.pyprecice.dist</li>
    <li>python312Packages.slepc4py</li>
    <li>python313Packages.petsc4py</li>
    <li>python313Packages.pyprecice</li>
    <li>python313Packages.pyprecice.dist</li>
    <li>python313Packages.slepc4py</li>
    <li>slepc</li>
  </ul>
</details>
